### ISSUE TITLE
LSP - Customize parentheses and quotes completion behavior

### ISSUE THREAD

**Issue by musjj** [godotengine/godot-proposals#6312]
### Describe the project you are working on

This proposal is project-agnostic, so I don't think a specific project context is necessary.

### Describe the problem or limitation you are having in your project

Users who use external editors like vim/neovim often have an autopairs plugin installed, and the habits tied with that plugin is usually deeply ingrained in their workflow.

For those who are not familiar with autopairs plugin, this is how it works:

![2023-02-18_08-25-28](https://user-images.githubusercontent.com/72612857/219824240-d650ec7e-dc74-49a9-9f9f-44ac567b7d9e.gif)
![2023-02-18_08-28-02](https://user-images.githubusercontent.com/72612857/219824323-84576b36-f648-4890-894b-31234e4f00e3.gif)

Basically, just as the name suggests, it autocompletes brackets, quotes or other types of pairs.

Now the problem is that Godot's LSP server does special things with brackets and quotes that breaks compatibility with this kind of plugins. While this behavior isn't _strictly_ specified by the LSP specification, no other LSP servers I've tried exhibits this kind of behavior, making it very unexpected and effectively non-standard.

Breaking expectations like this increases the difficulty of integrating new LSP servers when developing clients for other editors.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

There was a PR that attempted to deal with this issue by removing it: https://github.com/godotengine/godot/pull/49391
But it was rejected under the assertion that the feature _should_ be handled by the LSP.
Now, I agree that some user might find this behavior useful. So instead of removing it, why don't we expose an option to customize this behavior instead?

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

There's two things we can do:
- Expose this as an option somewhere in the editor settings.
- Expose this as a field in the `initializationOptions` that is sent during the initialization request to the server. Something like:
```lua
{
  gdscript = {
    auto_brackets = false
    auto_quotes = false
  }
}
```
Or maybe both.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No, this cannot be worked around with a script.

### Is there a reason why this should be core and not an add-on in the asset library?

I don't think this can be implemented as an add-on.

**Comment by musjj** [comment]
Some more comparisons
LuaLS:
![](https://i.imgur.com/oPnsMrj.gif)
![](https://i.imgur.com/FE7szxY.gif)
Godot:
![](https://i.imgur.com/TDx6tGK.gif)

**Comment by fjfnaranjo** [comment]
Maybe this is somehow related to https://github.com/godotengine/godot-proposals/issues/2215 .

Having the internal Godot editor using the internal LSP to implement code completion/aids (dogfooding) will make this strange interactions easy to spot during development.

Instead of adding configuration to the LSP server, having the parenthesis behavior isolated from the LSP looks more like current implementations work in other editors.

Following this pattern, the LSP can inform Godot's internal editor about the symbol nature (in this case, 'method') and the editor can add the parenthesis by itself. And then, the configuration will remain in the editor itself (something like an "Automatic parenthesis" setting).

What do you think?

While we wait for the LSP spec to advance and solidify (adding or discarding features like this), trying to follow the consensus of the most commonly used editors looks like a reasonable choice.

**Comment by musjj** [comment]
Yes, integrating the LSP into the built-in editor sounds like a great idea.
About the related issue, I guess this means Godot would have to either bundle other third-party LSPs (e.g. omnisharp, etc.) or rely on the one installed on the system. Once Godot have a general LSP client implemented, it shouldn't be too hard to add it.
So in the end what's the consensus on the current parentheses/quotes behavior? If it can just be safely removed, then that would also be a great option.
I don't use VSCode (I assume that's the editor of choice of most LSP consumers), but it probably have some kind of autopairing feature, just like neovim. So once it is implemented, they can just adapt the official VSCode LSP plugin accordingly (I've heard they had to do some extra workarounds due to this parentheses/quotes behavior?).

**Comment by fjfnaranjo** [comment]
> ... Godot would have to either bundle other third-party LSPs (e.g. omnisharp, etc.)

I can't think of any reason that makes reasonable to bundle arbitrary LSP server implementations with Godot.

Godot already uses the system .NET for C#. Looks like the only tools Godot should supply are the ones related to GDScript.

Godot dependencies are suppose to remain as minimal as possible.

> So in the end what's the consensus on the current parentheses/quotes behavior? If it can just be safely removed, then that would also be a great option.

The proposal is created and I guess you are free to explore the contribution guidelines to check how can you help to get it implemented as a LSP setting.

___

Said that... this issue points out to a complex situation that implies more than the issue itself:

- Whats the consensus about adding parenthesis in other LSP clients implementations?
- Should Godot follow it?
- Should all LSP clients add exceptions for Godot (I hope not)?
  - Is the setting a sufficient solution till we wait for the LSP community (other languages, server and client implementations) to settle in a particular strategy to handle adding parenthesis?

This requires coordination, research and consensus. It probably deserves another proposal and the attention of someone in the core team.

**Comment by Braveo** [comment]
I want to know if method autocomplete should also include parenthesis. I'm not sure if it's exactly related to this issue, but I am not a fan of `some_method(` being what's autocompleted. I wish it was simplified with `some_method` where you then have to manually type out the parenthesis. I'm not sure if there's one specific issue regarding this, but I wonder if it's related to this in some way.

I did my own little local branch to implement this:

https://github.com/user-attachments/assets/24eea5d6-7cff-4015-bf7f-19d7f09a8176

...but I wonder if it leads to undesirable effects for other text editors using the LSP.

**Comment by foob2483-oss** [comment]
@Braveo 
AFAIK this is just how every LSP under the sun works, regardless of the editor. Could you please make a PR out of your fork?

This issue has aged quite a bit and it's a shame that there are still no solutions to this day.

**Comment by HolonProduction** [comment]
I don't really see how the proposal is still relevant after https://github.com/godotengine/godot/pull/114598 and https://github.com/godotengine/godot/pull/117710

The main usecase outlined here is compat with autoclose plugins. This shouldn't be an issue anymore. We also aren't deviating from any standards. Inserting brackets with snippets is something other servers do as well (e.g. clangd).

**Comment by Braveo** [comment]
> I don't really see how the proposal is still relevant after [godotengine/godot#114598](https://github.com/godotengine/godot/pull/114598) and [godotengine/godot#117710](https://github.com/godotengine/godot/pull/117710)
> 
> The main usecase outlined here is compat with autoclose plugins. This shouldn't be an issue anymore. We also aren't deviating from any standards. Inserting brackets with snippets is something other servers do as well (e.g. clangd).

I think I may have misinterpreted this proposal. I think the parentheses completion behavior has been addressed already. I think the PR I opened was more related to customizing the opening of parentheses whenever a function autocompletes. I've still ran into the auto opening of parentheses in VSCode. Apologies for the confusion! I definitely want to discuss that further somewhere in a new proposal perhaps.

**Comment by HolonProduction** [comment]
Yeah I'd recommend you open a new proposal for your changes and their motivation. I don't think making a decision based on this proposal is possible, I'll probably going to close it at some point if no-one has any additional information about it.

**Comment by fjfnaranjo** [comment]
> Yeah I'd recommend you open a new proposal for your changes and their motivation. I don't think making a decision based on this proposal is possible, I'll probably going to close it at some point if no-one has any additional information about it.

Do you mean to close this current proposal (6312)?

**Comment by HolonProduction** [comment]
Yeah, the original use-case does not seem valid anymore, so I'm not confident the upvotes it gathered are still applicable to the current version. It would be better to track this in a new proposal that is just about making this configurable.

**Comment by fjfnaranjo** [comment]
As I see it, the OP and its initial post already cover this. It identifies the proposal nature an concludes that, looking at an already closed issue, the default behavior follows a particular intent and proposes a configuration option (see quote below). The issue is also titled correctly and the configuration would fit in the capabilities use case of the protocol.

> Describe the feature / enhancement and how it helps to overcome the problem or limitation
> There was a PR that attempted... ... it was rejected under the assertion that the feature should be handled by the LSP... ...we expose an option to customize this behavior instead?

If the team cannot verify that this configuration is needed anymore, you can request either the original user or another different one that relies on external editors to verify it, but if there is no reason to suspects it was resolved, I think "Good for concept/grab" fits better here.

Unless "Need further info" is currently being used as a "Will check it again later", something that doesn't looks unreasonable to me, but I cannot check by myself.

Do you think that is something that either OP or myself can help to support the continuation of the proposal?

**Comment by fjfnaranjo** [comment]
@HolonProduction I did some checking in advance of your reply and I cannot reproduce the behavior that motivated this proposal in the last Godot version (4.7.2).

We can ask OP to take a look (@musjj).

If this behavior is not present anymore, I suggest you close the PR instead of requesting more info.

**Comment by HolonProduction** [comment]
> Unless "Need further info" is currently being used as a "Will check it again later",

Yeah that's how I tend to use this stage. Proposals wouldn't go into the issue tracker otherwise anyway.

---

To cite from the OP:

> Users who use external editors like vim/neovim often have an autopairs plugin installed <...> Now the problem is that Godot's LSP server does special things with brackets and quotes that breaks compatibility with this kind of plugins.

If the premise under which a change was suggested has changed, this IMO needs a new proposal, since the community interest might have changed as well.

> but if there is no reason to suspects it was resolved

There is though, https://github.com/godotengine/godot/pull/114598 and https://github.com/godotengine/godot/pull/117710 ensure that you get matching quotes/brackets on insertion. This seems directly related to the problem statement.

Now while I use an external editor, I don't use an auto-pair plugin, which is why I left this open so that people can either confirm or deny my assumption about this working by now. If no-one does I'll assume I was right, that's just how we deal with inactive issues. And if an auto-pair user does confirm that there are still compat problems, we can obviously keep this open.

**Comment by fjfnaranjo** [comment]
Sorry, I wrote a second reply to my last comment at the same time you wrote this. You may want to read it.

**Comment by fjfnaranjo** [comment]
> Now while I use an external editor, I don't use an auto-pair plugin

Hum. I see. I never used an autopair before and I still tracked this issue.

Let me do a quick test with the 2 or 3 more common autopair plugins I can find and I will go back to you. We may be able to just close this.

**Comment by fjfnaranjo** [comment]
@HolonProduction 

I just tried my Neovim with both `windwp/nvim-autopairs` and `cohama/lexima.vim`.

Without the plugins, neither an isolated SH file nor a GD file completed quotes or parenthesis, as expected. The `.to_snake_case()` (no required args) was inserted atomically, also as expected. The Neovim->Godot LSP integration was able to complete without problems.

With any of the two plugins, closing quotes and parenthesis were inserted correctly and the snake method was also inserted without extra chars.

So, after that experiments, I still cannot reproduce the "special things with brackets and quotes".

Happy to run some more tests if you need them to make the call.

**Comment by HolonProduction** [comment]
I think that's sufficient ground to close the proposal. Thanks for testing!

Edit: To be clear, the proposal is being rejected because the described problem does not apply anymore. The change might still be desirable due to some different reason, but that requires a new proposal that makes the use-case clear.
